### PR TITLE
Issue 1238

### LINKED ISSUE #1238: Table character.account_tutorial not storing data
##### SMALL DESCRIPTION:

Characters created have to read tutorials all the times thei log in. AC not storing that they are already read.

##### EXPECTED BLIZZLIKE BEHAVIOUR:

Tutorials would be read just the first time the character log in.

##### CURRENT BEHAVIOUR:

When you login, all the times, you have to click to close all tutorials as the AC server don't store the info that they are already read in the DB table.

##### STEPS TO REPRODUCE THE PROBLEM:

1. Compile and start AC normally.
2. Login to the game and close all tutorials.
3. Close the game and login again.
4 . All tutorials reapeear.

##### EXTRA NOTES:

It works on Mangos and TrinityCore.

##### BRANCH(ES):

master

##### AC HASH/COMMIT:

commit aa949479d75f1060a622004ea75c2b15e3e999d7 (HEAD -> master, origin/master, origin/HEAD)

##### OPERATING SYSTEM:

Ubuntu 16.04

##### MODULES:

none

##### OTHER CUSTOMIZATIONS:

none

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/68324466-table-character-account_tutorial-not-storing-data?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Ercules76**
##### CHANGES PROPOSED:

-  Rewrote the logic to retrieve data from the DB

###### ISSUES ADDRESSED:

Closes #1238 

##### TESTS PERFORMED:

The build is without errors, tested in game after create a new account at lvl 1. Tested on windows 10

##### HOW TO TEST THE CHANGES:

- Close all tutorial panels, logout/exit game (the new row is in db) and no panels show on next login
- In settings panel disabled the "Show tutorial" and no panels show on next login
- When the tutorial is no longer shown, in settings panel reset the tutorial and the row in db is updated, panel are shown istant and persit if no closed

##### Target branch(es):

Master

### CODE REVIEW

File: src/server/game/Accounts/AccountMgr.cpp [hunk 1]
```diff
@@ -21,6 +21,9 @@ namespace AccountMgr
         if (utf8length(username) > MAX_ACCOUNT_STR)
             return AOR_NAME_TOO_LONG;                           // username's too long
 
+        if (utf8length(password) > MAX_PASS_STR)
+            return AccountOpResult::AOR_PASS_TOO_LONG;          // password's too long
+
```
Comment (Viste) [inline src/server/game/Accounts/AccountMgr.cpp]: How its related to the PR?

File: src/server/game/Accounts/AccountMgr.h [hunk 2]
```diff
@@ -21,6 +21,7 @@ enum AccountOpResult
 };
 
 #define MAX_ACCOUNT_STR 20
+#define MAX_PASS_STR 16
```
Comment (Viste) [inline src/server/game/Accounts/AccountMgr.h]: How its related to the PR?

File: src/server/game/Miscellaneous/Language.h [hunk 3]
```diff
@@ -876,8 +876,9 @@ enum TrinityStrings
     LANG_SQLDRIVER_QUERY_LOGGING_ENABLED  = 1027,
     LANG_SQLDRIVER_QUERY_LOGGING_DISABLED = 1028,
     // 1029-1030 used in other client versions
-
     // Room for more level 4              1031-1099 not used
+    LANG_ACCOUNT_PASS_TOO_LONG            = 1031,
+
```
Comment (Viste) [inline src/server/game/Miscellaneous/Language.h]: How its related to the PR?

File: src/server/scripts/Commands/cs_account.cpp [hunk 4]
```diff
@@ -113,6 +113,10 @@ class account_commandscript : public CommandScript
             handler->SendSysMessage(LANG_ACCOUNT_TOO_LONG);
             handler->SetSentErrorMessage(true);
             return false;
+        case AOR_PASS_TOO_LONG:
+            handler->SendSysMessage(LANG_ACCOUNT_PASS_TOO_LONG);
+            handler->SetSentErrorMessage(true);
+            return false;
```
Comment (Viste) [inline src/server/scripts/Commands/cs_account.cpp]: How its related to the PR?

File: data/sql/updates/pending_db_world/rev_1551539925032805900.sql [hunk 5]
```diff
@@ -0,0 +1,8 @@
+INSERT INTO `version_db_world` (`sql_rev`) VALUES ('1551539925032805900');
+
+DELETE FROM `trinity_string` WHERE `entry` = 1031;
+INSERT INTO `trinity_string` (`entry`, `content_default`) VALUES
+(1031, 'An account password can NOT be longer than 16 characters (client limit). Account NOT created.');
+
+UPDATE `trinity_string` SET `content_default` = 'Account name can\'t be longer than 20 characters (client limit), account not created!' WHERE (`entry` = '1005');
+
```
Comment (Viste) [inline data/sql/updates/pending_db_world/rev_1551539925032805900.sql]: How its related to the PR?

File: .gitignore [hunk 6]
```diff
@@ -82,12 +82,11 @@ local.properties
...
+#
 #  put below your custom ignore rules
-#  for example , if you want to include a 
+#  for example , if you want to include a
 #  module directly in repositoryyou can do:
 #
 #  !modules/yourmodule
 #
 # ==================
-
```
Comment (Viste) [inline .gitignore]: How its related to the PR?

File: .gitignore [hunk 7]
```diff
@@ -82,12 +82,11 @@ local.properties
 # ==================
 #
 #  CUSTOM
-#  
+#
 #  put below your custom ignore rules
-#  for example , if you want to include a 
+#  for example , if you want to include a
```
Comment (Viste) [inline .gitignore]: How its related to the PR?

File: src/server/game/Accounts/AccountMgr.cpp (hunk 1 above)
Comment (Ercules76) [reply src/server/game/Accounts/AccountMgr.cpp]: No it was my bad, i've created the PR with my fork on master branch, now this PR is merged than this snippet will not be considered.

File: src/server/game/Accounts/AccountMgr.h (hunk 2 above)
Comment (Ercules76) [reply src/server/game/Accounts/AccountMgr.h]: No it was my bad, i've created the PR with my fork on master branch, now this PR is merged than this snippet will not be considered.

File: src/server/game/Miscellaneous/Language.h (hunk 3 above)
Comment (Ercules76) [reply src/server/game/Miscellaneous/Language.h]: No it was my bad, i've created the PR with my fork on master branch, now this PR is merged than this snippet will not be considered.

File: src/server/scripts/Commands/cs_account.cpp (hunk 4 above)
Comment (Ercules76) [reply src/server/scripts/Commands/cs_account.cpp]: No it was my bad, i've created the PR with my fork on master branch, now this PR is merged than this snippet will not be considered.

File: data/sql/updates/pending_db_world/rev_1551539925032805900.sql (hunk 5 above)
Comment (Ercules76) [reply data/sql/updates/pending_db_world/rev_1551539925032805900.sql]: No it was my bad, i've created the PR with my fork on master branch, now this PR is merged than this snippet will not be considered.

**Comment by Viste** [comment]
@Ercules76 Can you reopen PR with right changes according only this problem.
Thanks.

**Comment by Ercules76** [comment]
> 
> 
> @Ercules76 Can you reopen PR with right changes according only this problem.
> Thanks.

yes, i think is better solutions

**Comment by BarbzYHOOL** [comment]
and what do you think about putting all the tutorial related stuff into the auth db, since it's account related?

**Pull request closed without merging**